File: problems/1012_numbers_with_repeated_digits/Solution.java
import java.util.*;

class Solution {
    public int numDupDigitsAtMostN(int n) {
        List<Integer> digits = new ArrayList<>();
        int temp = n + 1;
        while (temp > 0) { digits.add(0, temp % 10); temp /= 10; }
        int k = digits.size(), count = 0;
        for (int i = 1; i < k; i++) count += 9 * perm(9, i - 1);
        Set<Integer> seen = new HashSet<>();
        for (int i = 0; i < k; i++) {
            int d = digits.get(i);
            for (int x = (i == 0 ? 1 : 0); x < d; x++)
                if (!seen.contains(x)) count += perm(9 - i, k - i - 1);
            if (seen.contains(d)) break;
            seen.add(d);
        }
        return n - count;
    }
    private int perm(int n, int k) {
        int res = 1;
        for (int i = 0; i < k; i++) res *= (n - i);
        return res;
    }
}

File: problems/1012_numbers_with_repeated_digits/solution.py
class Solution:
    def numDupDigitsAtMostN(self, n: int) -> int:
        digits = [int(d) for d in str(n + 1)]
        k = len(digits)
        count = 0
        for i in range(1, k):
            count += 9 * self.perm(9, i - 1)
        seen = set()
        for i, d in enumerate(digits):
            for x in range(0 if i else 1, d):
                if x not in seen:
                    count += self.perm(9 - i, k - i - 1)
            if d in seen:
                break
            seen.add(d)
        return n - count

    def perm(self, n: int, k: int) -> int:
        result = 1
        for i in range(k):
            result *= (n - i)
        return result
